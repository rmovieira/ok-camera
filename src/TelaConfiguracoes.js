import React, { PureComponent } from "react";
import {
    Modal,
    StyleSheet,
    Text,
    TouchableHighlight,
    View,
    TextInput,
    Dimensions,
    Switch,
} from "react-native";
import Icon from 'react-native-vector-icons/Feather';



export default class TelaConfiguracoes extends PureComponent {

    state = {
        comandoFoto: 'Ok camera, tirar foto',
        comandoIniciarGravacao: 'Ok camera, iniciar gravação',
        comandoPararGravacao: 'Ok camera, parar gravação',
        habilitarVoz: this.props.habilitarVoz,
    }

    mudarComandoFoto = texto => {
        this.setState({ comandoFoto: texto });
    }

    mudarComandoIniciarGravacao = texto => {
        this.setState({ comandoIniciarGravacao: texto });
    }

    mudarComandoPararGravacao = texto => {
        this.setState({ comandoPararGravacao: texto });
    }

    fechar = () => {
        const valores = { ...this.state }
        this.props.fechar(valores);
    }

    habilitarVoz = valor => {
        this.setState({ habilitarVoz: valor });
    }

    render() {
        return (
            <Modal
                animationType="slide"
                transparent={true}
                visible={this.props.visivel}
                testID={'tela-configuracoes'}
            >
                <View style={styles.centeredView}>
                    <View style={styles.modalView}>
                        <Text testID={'titulo'} style={styles.titlo}>Configurações</Text>
                        <View style={styles.comandosVozContainer}>
                            <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
                                <Text style={styles.tituloComandosVoz}>Comandos de voz</Text>
                                <Switch
                                    testID={'habilitar-voz'}
                                    trackColor={{ false: "#767577", true: "rgba(153,0,0, 0.2)" }}
                                    thumbColor={this.state.habilitarVoz ? "#900" : "#f4f3f4"}
                                    onValueChange={this.habilitarVoz}
                                    value={this.state.habilitarVoz}
                                />
                            </View>
                            <View>
                                <View style={styles.comando}>
                                    <View style={styles.containerCaixaTexto}>
                                        <Icon name="camera" size={20} color="#900" style={styles.icone} />
                                        <TextInput
                                            testID={'comando-foto'}
                                            placeholder={'Exemplo: Tirar foto'}
                                            underlineColorAndroid={'black'}
                                            style={styles.caixaTexto}
                                            value={this.state.comandoFoto}
                                            maxLength={40}
                                            onChangeText={this.mudarComandoFoto}
                                            editable={this.state.habilitarVoz}
                                        />
                                    </View>
                                    <Text style={styles.contadorCaracteres} >  {`${this.state.comandoFoto.length}/40`}</Text>
                                </View>
                                <View style={styles.comando}>
                                    <View style={styles.containerCaixaTexto}>
                                        <Icon name="video" size={20} color="#900" style={styles.icone} />
                                        <TextInput
                                            testID={'comando-iniciar-gravacao'}
                                            placeholder={'Exemplo: Iniciar gravação'}
                                            underlineColorAndroid={'black'}
                                            style={styles.caixaTexto}
                                            value={this.state.comandoIniciarGravacao}
                                            maxLength={40}
                                            onChangeText={this.mudarComandoIniciarGravacao}
                                            editable={this.state.habilitarVoz}
                                        />
                                    </View>
                                    <Text style={styles.contadorCaracteres} >  {`${this.state.comandoIniciarGravacao.length}/40`}</Text>
                                </View>
                                <View style={styles.comando}>
                                    <View style={styles.containerCaixaTexto}>
                                        <Icon name="video-off" size={20} color="#900" style={styles.icone} />
                                        <TextInput
                                            testID={'comando-parar-gravacao'}
                                            placeholder={'Exemplo: Parar gravação'}
                                            underlineColorAndroid={'black'}
                                            style={styles.caixaTexto}
                                            value={this.state.comandoPararGravacao}
                                            maxLength={40}
                                            onChangeText={this.mudarComandoPararGravacao}
                                            editable={this.state.habilitarVoz}
                                        />
                                    </View>
                                    <Text style={styles.contadorCaracteres} > {`${this.state.comandoPararGravacao.length}/40`}</Text>
                                </View>
                                <View>
                                    <Text testID={'texto-explicativo'} style={styles.textoExplicativo}>Mesmo com o comando de voz habilitado, você pode realizar gravações e tirar fotos usando os botões que aparecem ao tocar na tela.</Text>
                                </View>
                            </View>
                        </View>
                        <TouchableHighlight
                            testID={'botao-fechar'}
                            style={styles.botaoFechar}
                            onPress={this.fechar}
                            underlayColor={'rgba(153,0,0, 0.2)'}
                            activeOpacity={.2}
                        >
                            <Text style={styles.textStyle}>Fechar</Text>
                        </TouchableHighlight>
                    </View>
                </View>
            </Modal>
        );
    }
}

const styles = StyleSheet.create({
    comandosVozContainer: { width: '100%', height: Dimensions.get('window').height - 200 },
    tituloComandosVoz: { fontSize: 20 },
    contadorCaracteres: { alignSelf: 'flex-end', color: 'gray', marginRight: 5 },
    icone: { alignSelf: 'center', marginRight: 10 },
    comando: { borderWidth: 0, width: '100%', },
    containerCaixaTexto: { width: '100%', flexDirection: 'row', height: 50 },
    caixaTexto: { flex: 1, fontSize: 20 },
    textoExplicativo: { color: '#505050', fontSize: 16, fontWeight: 'bold' },
    centeredView: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        marginTop: 0,
    },
    modalView: {
        margin: 20,
        backgroundColor: "white",
        borderRadius: 20,
        padding: 35,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2
        },
        shadowOpacity: 0.25,
        shadowRadius: 3.84,
        elevation: 5,
        flex: 1,
        height: '95%',
        width: '95%'
    },
    botaoFechar: {
        borderRadius: 20,
        padding: 10,
        backgroundColor: "#900",
    },
    textStyle: {
        color: "white",
        fontWeight: "bold",
        textAlign: "center"
    },
    titlo: {
        marginBottom: 15,
        textAlign: "center",
        fontSize: 25,

    }
});