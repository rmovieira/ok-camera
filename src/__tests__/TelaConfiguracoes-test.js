import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import TelaConfiguracoes from '../TelaConfiguracoes';

describe('Testes da tela de configurações', () => {

    beforeEach(() => {
        jest.useFakeTimers();
    });

    test('Deve montar a tela corretamente', async () => {
        const props = {
            visivel: true,
            habilitarVoz: false,
            fechar: () => { },
        }
        const tela = shallow(<TelaConfiguracoes {...props} />);

        expect(toJson(tela)).toMatchSnapshot();
    });

    test('Deve alterar os dados corretamente', async () => {
        const props = {
            visivel: true,
            habilitarVoz: true,
            fechar: () => { },
        }
        const tela = shallow(<TelaConfiguracoes {...props} />);

        expect(tela.find({ value: 'Ok camera, tirar foto' })).toExist();

        expect(tela.find({ value: 'Ok camera, iniciar gravação' })).toExist();

        expect(tela.find({ value: 'Ok camera, parar gravação' })).toExist();

        expect(tela.find({ testID: 'habilitar-voz' }).prop('value')).toBeTruthy();

        tela.find({ testID: 'comando-foto' }).simulate('changeText', 'texto do campo foto');
        expect(tela.find({ testID: 'comando-foto' }).prop('value')).toBe('texto do campo foto');

        tela.find({ testID: 'comando-iniciar-gravacao' }).simulate('changeText', 'texto do inicio de gravação');
        expect(tela.find({ testID: 'comando-iniciar-gravacao' }).prop('value')).toBe('texto do inicio de gravação');

        tela.find({ testID: 'comando-parar-gravacao' }).simulate('changeText', 'texto do fim da gravação');
        expect(tela.find({ testID: 'comando-parar-gravacao' }).prop('value')).toBe('texto do fim da gravação');

        tela.find({ testID: 'habilitar-voz' }).simulate('valueChange', false);

        expect(tela.find({ testID: 'habilitar-voz' }).prop('value')).toBeFalsy();
        expect(tela.find({ testID: 'comando-foto' }).prop('editable')).toBeFalsy();
        expect(tela.find({ testID: 'comando-iniciar-gravacao' }).prop('editable')).toBeFalsy();
        expect(tela.find({ testID: 'comando-parar-gravacao' }).prop('editable')).toBeFalsy();
    });

    test('Deve retornar os valores ao fechar a tela de configurações', async () => {
        const props = {
            visivel: true,
            habilitarVoz: false,
            fechar: jest.fn(),
        }
        const tela = shallow(<TelaConfiguracoes {...props} />);

        tela.find({ testID: 'comando-foto' }).simulate('changeText', 'texto do campo foto');
        expect(tela.find({ testID: 'comando-foto' }).prop('value')).toBe('texto do campo foto');

        tela.find({ testID: 'comando-iniciar-gravacao' }).simulate('changeText', 'texto do inicio de gravação');
        expect(tela.find({ testID: 'comando-iniciar-gravacao' }).prop('value')).toBe('texto do inicio de gravação');

        tela.find({ testID: 'comando-parar-gravacao' }).simulate('changeText', 'texto do fim da gravação');
        expect(tela.find({ testID: 'comando-parar-gravacao' }).prop('value')).toBe('texto do fim da gravação');

        tela.find({ testID: 'habilitar-voz' }).simulate('valueChange', true);

        tela.find({ testID: 'botao-fechar' }).props().onPress();

        const resultadoEsperado = {
            comandoFoto: 'texto do campo foto',
            comandoIniciarGravacao: 'texto do inicio de gravação',
            comandoPararGravacao: 'texto do fim da gravação',
            habilitarVoz: true,
        }
        expect(props.fechar).toHaveBeenCalledWith(resultadoEsperado);
    });

});