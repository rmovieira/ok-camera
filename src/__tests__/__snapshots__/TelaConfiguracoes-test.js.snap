// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes da tela de configurações Deve montar a tela corretamente 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  testID="tela-configuracoes"
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "flex": 1,
          "height": "95%",
          "margin": 20,
          "padding": 35,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "width": "95%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 25,
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
        testID="titulo"
      >
        Configurações
      </Text>
      <View
        style={
          Object {
            "height": 1134,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 20,
              }
            }
          >
            Comandos de voz
          </Text>
          <Switch
            onValueChange={[Function]}
            testID="habilitar-voz"
            thumbColor="#f4f3f4"
            trackColor={
              Object {
                "false": "#767577",
                "true": "rgba(153,0,0, 0.2)",
              }
            }
            value={false}
          />
        </View>
        <View>
          <View
            style={
              Object {
                "borderWidth": 0,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "height": 50,
                  "width": "100%",
                }
              }
            >
              <Icon
                allowFontScaling={false}
                color="#900"
                name="camera"
                size={20}
                style={
                  Object {
                    "alignSelf": "center",
                    "marginRight": 10,
                  }
                }
              />
              <Component
                allowFontScaling={true}
                editable={false}
                maxLength={40}
                onChangeText={[Function]}
                placeholder="Exemplo: Tirar foto"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flex": 1,
                    "fontSize": 20,
                  }
                }
                testID="comando-foto"
                underlineColorAndroid="black"
                value="Ok camera, tirar foto"
              />
            </View>
            <Text
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "gray",
                  "marginRight": 5,
                }
              }
            >
                
              21/40
            </Text>
          </View>
          <View
            style={
              Object {
                "borderWidth": 0,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "height": 50,
                  "width": "100%",
                }
              }
            >
              <Icon
                allowFontScaling={false}
                color="#900"
                name="video"
                size={20}
                style={
                  Object {
                    "alignSelf": "center",
                    "marginRight": 10,
                  }
                }
              />
              <Component
                allowFontScaling={true}
                editable={false}
                maxLength={40}
                onChangeText={[Function]}
                placeholder="Exemplo: Iniciar gravação"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flex": 1,
                    "fontSize": 20,
                  }
                }
                testID="comando-iniciar-gravacao"
                underlineColorAndroid="black"
                value="Ok camera, iniciar gravação"
              />
            </View>
            <Text
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "gray",
                  "marginRight": 5,
                }
              }
            >
                
              27/40
            </Text>
          </View>
          <View
            style={
              Object {
                "borderWidth": 0,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "height": 50,
                  "width": "100%",
                }
              }
            >
              <Icon
                allowFontScaling={false}
                color="#900"
                name="video-off"
                size={20}
                style={
                  Object {
                    "alignSelf": "center",
                    "marginRight": 10,
                  }
                }
              />
              <Component
                allowFontScaling={true}
                editable={false}
                maxLength={40}
                onChangeText={[Function]}
                placeholder="Exemplo: Parar gravação"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flex": 1,
                    "fontSize": 20,
                  }
                }
                testID="comando-parar-gravacao"
                underlineColorAndroid="black"
                value="Ok camera, parar gravação"
              />
            </View>
            <Text
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "gray",
                  "marginRight": 5,
                }
              }
            >
               
              25/40
            </Text>
          </View>
          <View>
            <Text
              style={
                Object {
                  "color": "#505050",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
              testID="texto-explicativo"
            >
              Mesmo com o comando de voz habilitado, você pode realizar gravações e tirar fotos usando os botões que aparecem ao tocar na tela.
            </Text>
          </View>
        </View>
      </View>
      <ForwardRef
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "backgroundColor": "#900",
            "borderRadius": 20,
            "padding": 10,
          }
        }
        testID="botao-fechar"
        underlayColor="rgba(153,0,0, 0.2)"
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Fechar
        </Text>
      </ForwardRef>
    </View>
  </View>
</Component>
`;
